### DISCUSSION TITLE
Customize the contacts-cache `SaveState` function.

### DISCUSSION THREAD

**Discussion by AndreaCatania** [jrouwe/JoltPhysics#611] [General]
Hello, I need to send the physics state, recorded using `SaveState`, from server to clients so that they can simulate the physics in a deterministic way.
**I need to optimize the data sent**, so I'm writing a custom `SaveState` that collects only the body state that the client is simulating. Notice, the server is simulating a large world and the clients are simulating only few bodies around the character.

So far this task seems doable, as the `SaveState` is already public, however the `Contact cache` is private and I'm wondering if I can PR a change that makes it public so I can write a custom `SaveState` that fetches only the contacts relative to the bodies simulated by the client.

Can I open a PR that allows that?

**Comment by jrouwe** [comment]
Hello,

The contact cache is something I'd rather not expose as it contains a lot of internal interface that users are not supposed to call.

I'm also not 100% sure if the simulation will be deterministic if you don't have the exact same bodies on both sides (it may be but is totally untested). In any case you need to be very careful in synching BodyIDs between simulation.

What about adding 2 virtual functions to StateRecorder that you can override? 

Something like:

```
bool ShouldSaveBody(const BodyID &)
```

and 

```
bool ShouldSaveConstraint(const Constraint &)
```

So that the contact cache (and body + constraint manager) can ask the recorder if the state of a particular body / constraint needs to be saved.

**Reply by AndreaCatania** [reply]
It would be super clean to handle in this way, so this is definitely a good approach 👌

Regarding the IDs,yes you are right there are still few things I'm trying to figure out but the idea is that the server propagates the IDs before applying the state.

Also, it would be interesting to see if the engine behaves deterministically even without networking all the contacts. I'll share this info with you once I test it.

Thx a ton for the help 🙏

**Reply by AndreaCatania** [reply]
I been reasoning about this and something else we need is to also improve the `RestoreState` so that it can restores the state for the bodies inside the received State.
In this way, I don't need to network all the static bodies as well. Do you think this is ok?

**Reply by jrouwe** [reply]
Looking at the code again, I think it's a non-trivial amount of work to get partial state saving to work. BodyManager, ConstraintManager and ContactConstraintManager are all not able to deal with partial/reordered state. I'm also not sure how it can be done in a way that doesn't affect performance too much. For things that are likely to get executed every frame, I avoid performance sensitive and memory hungry constructs like std::unordered_map and I also don't want any 3rd party dependencies. If you see a good way to fix this then propose the change in a PR.

**Reply by AndreaCatania** [reply]
Ok, it sounds a good plan (thx for not closing the door entirely). Can you please share the files::line where it's complex to operate with partial/reordered state restoration? It will simplify my job

**Reply by jrouwe** [reply]
The constraints array is assumed to be exactly the same as when saving:

https://github.com/jrouwe/JoltPhysics/blob/7f5d6eb618c701940dae8702caf20f34f841ceee/Jolt/Physics/Constraints/ConstraintManager.cpp#L242-L243

So a partial snapshot would restore settings on the wrong constraints.

The order of the constraint array matters because this makes the simulation deterministic:

https://github.com/jrouwe/JoltPhysics/blob/7f5d6eb618c701940dae8702caf20f34f841ceee/Jolt/Physics/Constraints/ConstraintManager.cpp#L106

Bodies are expected to be restored in the exact order as they were saved:

https://github.com/jrouwe/JoltPhysics/blob/7f5d6eb618c701940dae8702caf20f34f841ceee/Jolt/Physics/Body/BodyManager.cpp#L697-L702

Body ID consistency is required because otherwise this code will fail:

https://github.com/jrouwe/JoltPhysics/blob/7f5d6eb618c701940dae8702caf20f34f841ceee/Jolt/Physics/Constraints/ContactConstraintManager.cpp#L1342-L1360

And the active bodies are also problematic as it would overwrite everything (and thereby deactivate anything that's not part of the snapshot):

https://github.com/jrouwe/JoltPhysics/blob/7f5d6eb618c701940dae8702caf20f34f841ceee/Jolt/Physics/Body/BodyManager.cpp#L721-L729

And finally the ContactConstraintManager would only restore a partial contact cache:

https://github.com/jrouwe/JoltPhysics/blob/7f5d6eb618c701940dae8702caf20f34f841ceee/Jolt/Physics/Constraints/ContactConstraintManager.cpp#L452

Which would cause any bodies that were not part of the state to lose information about their previous contact and would cause issues when stacking multiple bodies.

**Reply by jrouwe** [reply]
What I think you're probably doing is to not call ```PhysicsSystem::SaveState``` at all, thereby avoiding most of the issues mentioned above as you are in control of calling ```Body::SaveState``` and ```Constraint::SaveState``` on the relevant bodies / constraints.

This means you're not executing this piece of code which is important if you have a variable time step / varying gravity:

https://github.com/jrouwe/JoltPhysics/blob/7f5d6eb618c701940dae8702caf20f34f841ceee/Jolt/Physics/PhysicsSystem.cpp#L2313-L2314

You would also be responsible for synching the active bodies. Because you don't execute this code:

https://github.com/jrouwe/JoltPhysics/blob/7f5d6eb618c701940dae8702caf20f34f841ceee/Jolt/Physics/Body/BodyManager.cpp#L721-L729

You would be responsible for ensuring that constraints are created in the same order because of what I mentioned before:

https://github.com/jrouwe/JoltPhysics/blob/7f5d6eb618c701940dae8702caf20f34f841ceee/Jolt/Physics/Constraints/ConstraintManager.cpp#L106

(very difficult because there is no API other than to remove / re-add constraints)

And as you say you're missing the saving of the contact cache:

https://github.com/jrouwe/JoltPhysics/blob/7f5d6eb618c701940dae8702caf20f34f841ceee/Jolt/Physics/Constraints/ContactConstraintManager.cpp#L1662

If there are no active bodies on the client that are not also active on the server then the way to fix this would be to add a callback function here to ask if a body pair needs saving or not (and expose the functionality through the PhysicsSystem).

I think this can work too, but is very fragile if I make changes to Jolt later and suddenly you're missing out on code that got added that you didn't know about.

**Reply by AndreaCatania** [reply]
> What I think you're probably doing is to not call `PhysicsSystem::SaveState`

Yes, TBH that was my initial idea, so I would be able to really customize what to sync.
Also, by splitting the sync state per body, would allow me to use the "delta update technique" by networking ONLY what changed: rather than sync everything each frame, that would result in a waste of bandwidth.

Luckily, there are just few things that are private and non accessible right now that are preventing me to collect the state manually, so change that will be "easy". I'll use this approach, and I'll PR any eventual new method added, to `Jolt`, to expose the internal stuff (like the contact cache, mention on the first comment).

**Reply by jrouwe** [reply]
This stayed in the back of my mind and I think there is an elegant way of solving this:

https://github.com/jrouwe/JoltPhysics/tree/feature/selective_state_saving

Note that I did not validate that a partial state save actually works (conceptually it should but there may be bugs).

**Reply by AndreaCatania** [reply]
Cool I'll test this as soon as I'm at work 👌

**Reply by jrouwe** [reply]
I pushed another update because what I had written had a fundamental issue: Your clients don't have the same set of constraints as the server, the constraint serialization wouldn't have worked and it wouldn't have been deterministic either.

The state recorder filter now allows you to turn off saving of constraints, which you're going to need to do (there is no identifier like BodyID that is consistent over the network). You then need to write custom code to save and restore their state by directly calling Constraint::SaveState/RestoreState. I added a Constraint::SetConstraintPriority function that will need a unique number for every constraint. This number needs to be the same on client and server. This makes it possible to add/remove constraints in any desired order and to have a different set of constraints on client and server and still maintain determinism.

As before: This is totally untested code (I only tested it without filter).

I'm aware of one more issue and that is that any body that is active on the client and that is not part of the restore, will be deactivated. This can be fixed if you need it to, but it needs a little bit more work.

**Reply by AndreaCatania** [reply]
This looks promising, thanks a lot, I'll test this with my test scene and share results. Eventually I can write some tests.

> I'm aware of one more issue and that is that any body that is active on the client and that is not part of the restore, will be deactivated. This can be fixed if you need it to, but it needs a little bit more work.

Yes, this is perfect I think; the received state will contains all the active bodies anyway. The fact that it deactivates the bodies that are not contained into the Server state, is a really good feature as I don't want that the untracked bodies are being active on the client.

**Reply by jrouwe** [reply]
I've merged Constraint::SetConstraintPriority into master in #638, the other changes are still in https://github.com/jrouwe/JoltPhysics/tree/feature/selective_state_saving waiting for me to have time to test them properly.

**Reply by AndreaCatania** [reply]
Ok nice thanks for the heads-up, I'm fully focused on this task now. I'll be able to provide additional info during the next days.

**Reply by AndreaCatania** [reply]
@jrouwe Do you have somewhere a clang-formatter file with Jolt formatting settings?
It will simplify my job; as on save, the jolt files are being entirely formatted following another project format, making it difficult to work with.

**Reply by jrouwe** [reply]
No, I haven't been able to make a clang-format file that matches the style of the code, sorry.

**Reply by AndreaCatania** [reply]
Ok no problem, thanks.

**Reply by AndreaCatania** [reply]
I been testing this while creating this unit test: https://github.com/jrouwe/JoltPhysics/pull/643 and everything seems to work well.

However, I been reasoning about this:
> I'm aware of one more issue and that is that any body that is active on the client and that is not part of the restore, will be deactivated. This can be fixed if you need it to, but it needs a little bit more work.

~~All the "unsync" bodies, that now get automatically deactivated, will see their transform changed overtime (due to some interpolation that the server is going to stream). I don't need any collision detection, nor I'll use `KinematicMove`, I'll just update the transform. I don't think this is going to create any issue, though I'm wondering if this sounds ok to you as well. Keep in mind that these unsync bodies, will be faraway from the client collider and there is no way for those to affect the client active bodies.~~

Never mind, I've just expanded the UnitTest, to verify that:
- Teleporting deactivated bodies doesn't create any issue.
- Validate body activation / deactivation due to State Save / Restore.

**Reply by jrouwe** [reply]
I simplified the interface and updated the unit test to include partial activation/deactivation too. The change has been merged into the master branch.

**Reply by AndreaCatania** [reply]
Nice, thanks a lot. I been checking the new API and I like it, however something that I don't like as much is that the `EStateRecorderState` is being saved into the stream. I think that it should be an optional parameter of `RestoreState(StateRecorder &inStream, EStateRecorderState inState = EStateRecorderState::All)`, because usually both the sender and receiver knows how to apply the state and if they don't know it's not that hard, for the user, encode that information into the buffer.

I know that 8 bits doesn't sound like a lot, but sending this state each 3/4 frames is going to be expensive, so I'm trying to save as much as I can by lowering the amount repeated data. For example, I'm re-creating the body manager save state to implement the delta update.

Do you think that would be ok making `EStateRecorderState` a paramter of `RestoreState` instead?
In any case, thanks for the help!

**Reply by AndreaCatania** [reply]
Btw, I love the way the activation status is now encoded. It doesn't require a quick sort, and the code is simpler.

**Reply by jrouwe** [reply]
> Do you think that would be ok making EStateRecorderState a paramter of RestoreState instead?

I had that at first but it results in crashes if you pass the wrong value (which is very easy because the parameter has a default value) so I made it into what it is now. For context, saving a single dynamic body uses 142 bytes in single precision mode and more in double precision mode, so I don't think that byte is too much data.

**Reply by AndreaCatania** [reply]
Yeah, make sense, my idea is to reduce body size as much as I can, using delta update (and keeping ""static"" parameters, like the friction, mass out), by implementing a custom `SaveState`. I will use the default `SaveState` + the filter, just to store the needed contacts.
At some point I want to do something about BodyID too, as I think we can shrink it even more. Eventually, I'll figure out how to keep that byte out as well. In any case, I'll share with you some ideas, maybe we can figure out a way to make `SaveState` really customizable and granular while keeping the API simple.

**Reply by jrouwe** [reply]
> Yeah, make sense, my idea is to reduce body size as much as I can, using delta update (and keeping ""static"" parameters, like the friction, mass out), by implementing a custom SaveState. I will use the default SaveState + the filter, just to store the needed contacts.

Actually it was always the intention that only the state that got modified by the simulation was saved (constraints already do) but it appears that too much body state was saved. I've reduced the amount of state that gets saved to that what is simulated in #654.

**Reply by AndreaCatania** [reply]
Oh  wow, this is going to be useful, thanks a lot!

I was wondering about this:
```
inStream.Write(mAllowSleeping);
```

Is this going to change at runtime because of the Physics Simulation? If not, we should consider to keep it out, and leave to the user the responsibility to network it.

**Reply by jrouwe** [reply]
It's unfortunately needed because the vehicles switch allow sleeping mode at run time
